### PR TITLE
Html helper

### PR DESCRIPTION
**Pull request by katejeffreys**
Staged here: 

https://html-helper-dot-polymer-project.appspot.com/2.0/docs/devguide/dom-template

### CODE REVIEW

**Comment by katejeffreys** [comment]
Fixes #2407

**Comment by justinfagnani** [comment]
LGTM!

My only small nit is not with this change, but how we talk about properties like `template` and `properties`.  We always call them the "static X getter", but really, a getter is just one way to implement the property.

You could also just set the property on the constructor:

```js
class MyElement {}
MyElement.template = html`...`;
```

Or if you using a compiler, and in the near future in native JS, declare a static property:

```js
class MyElement {
  static template = html`...`;
}
```

**Comment by katejeffreys** [comment]
This is a great point & something I was processing this morning after
talking to Arthur about templates in a static context.

I created an issue in the docs to consider more general ways of referring
to `template` and `properties`. #2414

Is it okay to merge now, and address this later? It goes deeper than this
issue, as you said, and could be addressed throughout the site.

**Comment by justinfagnani** [comment]
Oh yeah, this looks good to me as-is, I just wanted to mention that nit since it's been on my mind for a while.

**Comment by katejeffreys** [comment]
Sweet, thanks Justin. I'll get Arthur to give it a look, then merge if he's cool with it.

File: app/2.0/docs/devguide/dom-template.md [hunk 1]
```diff
@@ -14,18 +14,18 @@ The DOM template is also processed to enable data binding and declarative event
...
--   `<dom-module>` element. This allows you to specify the template entirely in markup,
-    which is most efficient for an element defined in an HTML import.
--   Provide a string template. This works well for elements defined entirely in JavaScript
-    (for example, in an ES6 module).
--   Retrieve or generate your own template element. This allows you to define how the template is
-    retrieved or constructed, and can be used to modify a superclass template.
+-   [Specify a template using the `<dom-module>` element](#dommodule). This allows you to specify the
+    template entirely in markup, which is most efficient for an element defined in an HTML
+    import.
+-   [Assign a string template to an HTMLTemplateElement object, and return the
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: I like the new links.

I think the description of this one focuses too much on the implementation. The gist here is that you can define a `template` getter that returns the template. 

Except that @justinfagnani raised a concern about the wording of that. Um, so,...

"Add a template property on the constructor that returns the template?"

"Provide a template by defining a template property on the constructor"?

File: app/2.0/docs/devguide/dom-template.md [hunk 2]
```diff
@@ -60,106 +60,184 @@ except for testing. For caveats about defining elements in the main document,
...
-       <div>I've got a string template!</div>`
+       <div>I've got a string template!</div>`;
   }
 }
 customElements.define('my-element', MyElement);
 ```
 
-When using a string template, the element doesn't need to provide an `is` getter (however, the tag
-name still needs to  be  passed as the first argument to `customElements.define`).
+Alternatively, declare a variable called `html` to hold the `Polymer.html` object, and
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: I feel like this should be listed goal-first. It's a lot of text to read through to figure out that you'd want to do this to enable syntax highlighting.

File: app/2.0/docs/devguide/dom-template.md [hunk 3]
```diff
@@ -60,106 +60,184 @@ except for testing. For caveats about defining elements in the main document,
 see [main document definitions](registering-elements#main-document-definitions).
 {.alert .alert-info}
 
-### Specify a string template
+### Return an HTMLTemplateElement {#templateobject}
 
-As an alternative to specifying the element's template in markup, you can specify a string template
-by creating a static `template` getter that returns a string.
+As an alternative to specifying the element's template in markup, you can create a static
+`template` getter that returns an HTMLTemplateElement.
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: HTMLTemplateElement should be in code format.

The magic `Polymer.html` should be introduced, like,  "Polymer provides an `html` function that can be applied to a template literal to produce an `HTMLTemplateElement`." Or something @justinfagnani ? 

Unfortunately, ECMAScript kind of failed when it comes to terminology here. `foo` is a template literal. html`foo` is a _tagged template literal_. What exactly the `html` function is in this syntax is not specified, AFAICT, in the ECMAScript spec. It's... the function that you tag the literal with. Or something.

Sometimes I hate engineers. Just sayin'.

Anyway, point is, I'm basically just specifying a string, but presto change-o, I return a full-blown template element.

**Comment by arthurevans** [comment]
Few comments in today. I have to run, will finish up tomorrow. I think we should iterate a little more here...

**Comment by katejeffreys** [comment]
@arthurevans Great points, thank you. Incorporated all. Let's iterate until it's ready~

File: app/2.0/docs/devguide/dom-template.md [hunk 4]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
...
+* To [**extend** a base class template](#extendbase), supply a template in the child class that includes
+  a reference to the base class template.
+* To [**insert** content from a child class into placeholders supplied by a base class](#insertcontent),
+  create placeholders in the base class and supply the content with functions.
 
--   Don't modify the superclass template in place; make a copy before modifying.
--   If you're doing anything expensive, like copying or modifying an existing template,
-    you should memoize the modified template so you don't have to regenerate it when the
-    getter is called.
+#### Inherit a base class template with no modifications {#nomods}
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: These examples should use dom-module, since it's the default mechanism for Polymer 2.x.

**Comment by katejeffreys** [comment]
http://plnkr.co/edit/f5utiX?p=preview does not work

**Comment by katejeffreys** [comment]
Neither does http://plnkr.co/edit/YQ5t5I?p=preview

File: app/2.0/docs/devguide/dom-template.md (hunk 4 above)
Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: Converted back to dom-module examples - a couple of issues with the following

http://plnkr.co/edit/f5utiX?p=preview
http://plnkr.co/edit/YQ5t5I?p=preview

@arthurevans @justinfagnani

File: app/2.0/docs/devguide/dom-template.md [hunk 5]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
 see [main document definitions](registering-elements#main-document-definitions).
 {.alert .alert-info}
 
-### Specify a string template
+### Provide a template by defining a template property on the constructor {#templateobject}
 
-As an alternative to specifying the element's template in markup, you can specify a string template
-by creating a static `template` getter that returns a string.
+As an alternative to specifying the element's template in markup, you can define a static `template`
+property. The `template` property must be an `HTMLTemplateElement`.
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: must be => should be

This section is correct for 3.x, but for 2.x we support returning a string, as well.

Maybe in place of the second sentence, add something to the next para, like,

In Polymer 2.x, the template can be supplied as an `HTMLTemplateElement` instance or a string. Support for returning a string template is being discontinued in Polymer 3.x. Starting in Polymer 2.4 <<<verify actual release when it lands>>>, Polymer provides ... This is the recommended way of defining a string template in Polymer 2.4 and later.

File: app/2.0/docs/devguide/dom-template.md [hunk 6]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
...
-       <div>I've got a string template!</div>`
+       <div>I've got a string template!</div>`;
   }
 }
 customElements.define('my-element', MyElement);
 ```
 
-When using a string template, the element doesn't need to provide an `is` getter (however, the tag
-name still needs to  be  passed as the first argument to `customElements.define`).
+To enable HTML code highlighting in text editors with lit-html code highlighting functionality, you
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: "with lit-html code highlighting functionality" => I gather this isn't strictly true; some just have support for this built-in according to @justinfagnani ... Also, somewhat confusing to mention this in this context, because the `html` already _looks_ kind of like lit-html, and we don't want people to confuse the two.

"some text editors" is kind of lame. 

the full explanation is probably a bit wordy, would probably be something like,
"Some text editors support HTML code highlighting in template literals tagged with a function named `html`." Again, @justinfagnani to confirm perhaps?

File: app/2.0/docs/devguide/dom-template.md [hunk 7]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
...
+can declare a variable called `html` to hold the `Polymer.html` object, and return it from a `template` getter:
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: Two details: `Polymer.html` is a function, and `html` here is a constant. 

"And return it from the template getter..." sounds like you're returning the function. Use it in the template getter, perhaps?

File: app/2.0/docs/devguide/dom-template.md [hunk 8]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
...
+  static get template() {
+    return html`
+      This content has been inherited from BaseClass.
+      `;
+  }
+}
+```
+
+Child class definition { .caption }
+```js
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: Arguably this is simpler than using the existing template manipulation code, but it involves some magic that should probably be explained. 

1) The `${expression}` inside a template literal interpolates the value of an expression. (This is basic ES6, and maybe it doesn't need to be spelled out, but if so, we should at least link to the MDN description of template literals at the first reference (in the string template section), because we're using a whole bunch of their features here and not everyone is familiar with them yet.)
2)  Because the getter is static, `super.template` accesses the template property on the superclass constructor.
3) The `Polymer.html` function checks each expression value. If the value is a string, the string is interpolated directly. If the expression is an `HTMLTemplateElement`, the `innerHTML` of the template is interpolated. This makes it very simple to combine existing templates.

Comment (arthurevans) [reply app/2.0/docs/devguide/dom-template.md]: And this is a lot of words, and maybe should be pushed up to some concepts section somewhere, but I think it's important because we're using a lot of ES6-fu here that seems completely transparent to the @justinfagnani s of the world...

File: app/2.0/docs/devguide/dom-template.md [hunk 9]
```diff
@@ -60,106 +59,183 @@ except for testing. For caveats about defining elements in the main document,
...
+    ${super.template}
+    <p>Hello again from ChildClass.</p>
+    `;
+  }
+}
 ```
 
+[See a working example in Plunker](http://plnkr.co/edit/f5utiX?p=preview).
+
+#### Insert content from a child class into placeholders supplied by a base class {#insertcontent}
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: I'm not sure this is the correct title for this use case... It seems like the primary use case here is the _superclass_ use case: that is, "provide overrideable sub-templates." Or something.  Replaceable sub-templates? Allow easy template extension with modular templates?

@justinfagnani @kevinpschaaf @graynorton Anyone want to name this pattern? :D 

Just seems to me that the important part is, a superclass providing a focused way to extend its template with content. They could be placeholders, or not—for example, they could be a valid part of the superclass template that a subclass may or may not replace.

Comment (arthurevans) [reply app/2.0/docs/devguide/dom-template.md]: Also: We may want to leave in one or both of the old-school template extension examples, since they're still valid and ```html`${foo}` ``` will only work in 2.4 and later.

File: app/2.0/docs/devguide/dom-template.md (hunk 4 above)
Comment (arthurevans) [reply app/2.0/docs/devguide/dom-template.md]: I meant that specifically for the "inherit a base class with no modifications."

The `${expr}` format is specific to template literals (those things in ``` `backticks` ```) and won't work in `<template>` elements... despite the names, they're rather different creatures.

The extension examples you added with the `html` function and template literals are much clearer than the previous ones that require manipulating template elements in JS. And for that reason I think we might want to keep them in. However, they'll only work in 2.4+, so we might want to leave in at least one of the previous examples.

I had some more specific comments on this stuff inline.

File: app/2.0/docs/devguide/dom-template.md (hunk 9 above)
Comment (kevinpschaaf) [reply app/2.0/docs/devguide/dom-template.md]: * "Template extension points"
* "Overridable sub-templates"

I'd probably go with the first...?

**Comment by katejeffreys** [comment]
@arthurevans @justinfagnani Revised the page based on feedback. Please have another look. https://www.polymer-project.org/2.0/docs/devguide/dom-template is updated with the revised version.

**Comment by justinfagnani** [comment]
@katejeffreys the one question I have is whether or not it's confusing to mention plain string templates at all. The "Specify a string template in a static template getter" example will be deprecated in the next release. Maybe remove that example and add a note that says something like "Note: It used to be possible to specify a plain string template as a static property on your element class, but this feature was deprecated in 2.? and will be removed in Polymer 3.0"

File: app/2.0/docs/devguide/dom-template.md [hunk 10]
```diff
@@ -76,19 +91,148 @@ class MyElement extends Polymer.Element {
...
+customElements.define('my-element', MyElement);
+```
+
+### Inherited templates {#inherit}
 
 An element that extends another Polymer element can inherit its template. If the element doesn't
-provide its own DOM template (using either a `<dom-module>` or a string template), Polymer uses the
-same template as the superclass, if any.
+provide its own DOM template (using either a `<dom-module>` or a static template object), Polymer
+uses the same template as the superclass, if any.
```
Comment (arthurevans) [inline app/2.0/docs/devguide/dom-template.md]: I think this list of items needs some introduction, like: 

The `Polymer.html` helper introduced in Polymer 2.4 makes template extension a little simpler, so this section shows examples with `Polymer.html` (usable on 2.4 and later) and without (usable on all 2.x versions).

Or... something.

**Comment by arthurevans** [comment]
Ping @katejeffreys on the updates here. As soon as we have those updates we can merge this for the 2.4 release and cut a 2.x branch.

**Comment by katejeffreys** [comment]
Thanks for the reminder! Will do.

File: app/2.0/docs/devguide/dom-template.md [hunk 11]
```diff
@@ -4,28 +4,29 @@ title: DOM templating
...
 
 Polymer provides three basic ways to specify a DOM template:
 
--   `<dom-module>` element. This allows you to specify the template entirely in markup,
-    which is most efficient for an element defined in an HTML import.
--   Provide a string template. This works well for elements defined entirely in JavaScript
-    (for example, in an ES6 module).
--   Retrieve or generate your own template element. This allows you to define how the template is
-    retrieved or constructed, and can be used to modify a superclass template.
+-   [Specify a template using the `<dom-module>` element](#dommodule). This allows you to specify
```
[the remark this answers is not in the document]
Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: - [ ] update headings and links to make parallel & simpler

File: app/2.0/docs/devguide/dom-template.md [hunk 12]
```diff
@@ -53,19 +54,33 @@ Example: { .caption }
...
 **Note:** Elements should generally be defined outside of the main document,
 except for testing. For caveats about defining elements in the main document,
 see [main document definitions](registering-elements#main-document-definitions).
 {.alert .alert-info}
 
-### Specify a string template
+### Provide a template by defining a template property on the constructor {#templateobject}
+
+As an alternative to specifying the element's template in markup, you can define a static 
+`template` property.
```
[the remark this answers is not in the document]
Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: - [ ] static template property => template property on the constructor.

File: app/2.0/docs/devguide/dom-template.md [hunk 13]
```diff
@@ -53,19 +54,33 @@ Example: { .caption }
...
 {.alert .alert-info}
 
-### Specify a string template
+### Provide a template by defining a template property on the constructor {#templateobject}
+
+As an alternative to specifying the element's template in markup, you can define a static 
+`template` property. 
+
+One way to do this is to specify a string template by creating a static template getter that 
+returns a template literal.
```
[the remark this answers is not in the document]
Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: - [ ] clarify intro to this section
- [ ] make sure terminology is correct
- [ ] subsection: 2.4+ 
- [ ] subsection: 2.x all

**Comment by justinfagnani** [comment]
@katejeffreys did you see https://github.com/Polymer/polymer/pull/5023

We're not going to allow arbitrary values for interpolations into templates. Only other templates.

**Comment by katejeffreys** [comment]
Thx for heads-up @justinfagnani. Editing now.

File: app/2.0/docs/devguide/dom-template.md (hunk 10 above)
Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: - [ ] intro to list

**Comment by katejeffreys** [comment]
@arthurevans @justinfagnani This is ready for another look.

File: app/2.0/docs/devguide/dom-template.md [hunk 14]
```diff
@@ -65,42 +61,32 @@ except for testing. For caveats about defining elements in the main document,
...
+    // Polymer 2.4+: Return HTMLTemplateElement 
   }
 }
-
-customElements.define('my-element', MyElement);
 ```
 
-[See a working example in Plunker](http://plnkr.co/edit/AzkMDJ?p=preview).
+Polymer 2.4+ provides a helper function (`Polymer.html`) to generate an `HTMLTemplateElement`
+instance from a template literal.
```
Comment (justinfagnani) [inline app/2.0/docs/devguide/dom-template.md]: I'd do "JavaScript template literal" to be extra clear, since we're also talking about HTML templates too

File: app/2.0/docs/devguide/dom-template.md [hunk 15]
```diff
@@ -141,12 +127,41 @@ class MyElement extends Polymer.Element {
...
+#### Return a string template (all 2.x versions) {#stringtemplate}
+Template literals are denoted by being enclosed in back-ticks (<code>` `</code>). 
+
+Specify a string template in a static template getter: { .caption }
+
+```js
+class MyElement extends Polymer.Element {
+
+  static get template() {
+    return `<style>:host { color: blue; }</style>
```
Comment (justinfagnani) [inline app/2.0/docs/devguide/dom-template.md]: Doesn't this need to be:

```js
  return html`<style>:host { color: blue; }</style>
```

Comment (katejeffreys) [reply app/2.0/docs/devguide/dom-template.md]: @justinfagnani per @arthurevans' feedback, I left that example in for users of pre-2.4 versions. 

I may have misunderstood which forms are available in which versions, though. My understanding was that in 2.4+, the string template return value is deprecated but still usable. For that reason the heading indicates that the example applies to "all 2.x versions" - would you mind clarifying on that point?

@arthurevans WDYT?

**Comment by katejeffreys** [comment]
This is almost ready to merge, I think. I need a couple points of clarification though:

* Which versions the string template is still a possible return value of `static get template(){...}` (i.e. 2.x or only < 2.4) 
* Whether to include the string template example in the docs (my preference is to keep it)

@arthurevans @justinfagnani

**Comment by kevinpschaaf** [comment]
@katejeffreys 

> * Is a string template still a possible return value of `static get template(){...}` in 2.4+?

It is deprecated (a warning will print), but still allowed.  The intention is to actually remove support before the final 3.0 release is tagged.

> * Should we keep the string template example in the docs for users of pre-2.4 versions?

I'd say no, let's switch to using `Polymer.html` (maybe with a note that this is available in 2.4)?

**Comment by arthurevans** [comment]
@kevinpschaaf @katejeffreys For the string template, I think we should avoid giving people the impression that they need to change this now if we're going to provide a we're going to provide an automated transition to 3.0. Plus, some people won't upgrade to 2.4 right away. So we should definitely provide the alternative.

How about we show the Polymer.html version first, then note: `Polymer.html` was added in release 2.4. For earlier versions, you can return a plain string:

```
    return `<div>A plain string template</div>`;
```

Returning a string is deprecated in favor of the `html` helper in release 2.4. Support for the string template will be removed in 3.0.

... Or words to that effect.

**Comment by katejeffreys** [comment]
@arthurevans @kevinpschaaf @justinfagnani Updated and staged here: https://html-helper-dot-polymer-project.appspot.com/2.0/docs/devguide/dom-template#templateobject
